pipelines/whale/extractor/cloud_spanner_extractor.py

```diff
@@ -0,0 +1,129 @@
+      lower(c.table_schema) AS `schema`,
+      lower(c.table_name) AS name
+    FROM
+      INFORMATION_SCHEMA.COLUMNS AS c
+    LEFT JOIN
+      INFORMATION_SCHEMA.TABLES t
+          ON c.TABLE_NAME = t.TABLE_NAME
+          AND c.TABLE_SCHEMA = t.TABLE_SCHEMA
+    {where_clause_suffix}
+    ORDER by `schema`, name, col_sort_order ;
+    """
+
+    HEADER = ["col_name", "col_type", "col_sort_order", "schema", "name"]
+
+    # Config keys.
+    DEFAULT_CONFIG = ConfigFactory.from_dict({
+        WHERE_CLAUSE_SUFFIX_KEY: '',
+    })
+
+    def init(self, conf):
+        self.conf = conf.with_fallback(CloudSpannerExtractor.DEFAULT_CONFIG)
+        self._instance_id = self.conf.get_string(CloudSpannerExtractor.INSTANCE_ID_KEY)
+        self._database_id = self.conf.get(CloudSpannerExtractor.DATABASE_ID_KEY, None)
+        self._key_path = self.conf.get(CloudSpannerExtractor.KEY_PATH_KEY, None)
+
+        client_kwargs = {
+            "project": self.conf.get(CloudSpannerExtractor.PROJECT_ID_KEY)
+        }
+        if self._key_path is not None:
+            spanner_client = spanner.Client.from_service_account_json(self._key_path, **client_kwargs)
+        else:
+            spanner_client = spanner.Client(**client_kwargs)
+
+        self.instance = spanner_client.instance(self._instance_id)
+        self.database = self.instance.database(self._database_id)
+
+        self.sql_stmt = CloudSpannerExtractor.SQL_STATEMENT.format(
+            where_clause_suffix=self.conf.get_string(
+                CloudSpannerExtractor.WHERE_CLAUSE_SUFFIX_KEY)
+        )
+
+        LOGGER.info('SQL for cloud spanner: {}'.format(self.sql_stmt))
+
+        self._extract_iter = None  # type: Union[None, Iterator]
+
+    def extract(self):
+        # type: () -> Union[TableMetadata, None]
+        if not self._extract_iter:
+            self._extract_iter = self._get_extract_iter()
+        try:
+            return next(self._extract_iter)
+        except StopIteration:
+            return None
+
+    def get_scope(self):
+        # type: () -> str
+        return 'extractor.presto_table_metadata'
+
+    def _get_extract_iter(self):
+        # type: () -> Iterator[TableMetadata]
+        """
+        Using itertools.groupby and raw level iterator, it groups to table and yields TableMetadata
+        :return:
+        """
+
+        with self.database.snapshot() as snapshot:
+            results = snapshot.execute_sql(self.sql_stmt)
+            header = CloudSpannerExtractor.HEADER
+            headered_results = [dict(zip(header, result)) for result in results]
+
+            for _, group in groupby(headered_results, self._get_table_key):
+                columns = []
+
+                for row in group:
+                    last_row = row
+                    columns.append(ColumnMetadata(row['col_name'], None,
+                                                  row['col_type'], row['col_sort_order']))
+
+            yield TableMetadata(self._instance_id,
```

Comment:
@rsyi Ran it locally, I think `yield...` is missing one indent (yield for each table, not only once).

Reply:
amazing catch, thanks @manuelzander!